### PR TITLE
cd_adf: Added importance weight probability clipping for cb_type mtr, ips, dr

### PR DESCRIPTION
**Pull request by marco-rossi29**

### CODE REVIEW

**Comment by JohnLangford** [comment]
What's the goal here?

**Comment by marco-rossi29** [comment]
To be able to clip the important weight in MTR and to be able to use it as a cmd arg when needed.

**Comment by JohnLangford** [comment]
Why limit this to only mtr?

**Comment by marco-rossi29** [comment]
Only implementaton time reasons, but I'm happy to expand to any cb_type as soon as possibile. Implementing it for dr and ips was much more involved than mtr since for dr and ips the probability is used outside of `cd_adf.cc`

**Comment by JohnLangford** [comment]
It seems better to have a relatively universal implementation in the master branch?  Maybe we can discuss what's needed next week.

**Comment by marco-rossi29** [comment]
Sounds good!

File: vowpalwabbit/gen_cs_example.h [hunk 1]
```diff
@@ -11,6 +11,8 @@
 #include "cb_algs.h"
 #include "vw_exception.h"
 
+using namespace std;
```
Comment (lokitoth) [inline vowpalwabbit/gen_cs_example.h]: It looks like you are using explicit std::max throughout your code additions. Is the using still necessary? If not, prefer to remove it.

File: vowpalwabbit/gen_cs_example.h [hunk 2]
```diff
@@ -137,14 +139,14 @@ void gen_cs_label(cb_to_cs& c, example& ec, COST_SENSITIVE::label& cs_ld, uint32
         ((c.known_cost->cost - wc.x) * (c.known_cost->cost - wc.x) - c.avg_loss_regressors);
     c.last_pred_reg = wc.x;
     c.last_correct_cost = c.known_cost->cost;
-    wc.x += (c.known_cost->cost - wc.x) / c.known_cost->probability;
+    wc.x += (c.known_cost->cost - wc.x) / (std::max)(c.known_cost->probability, clip_p);
```
Comment (lokitoth) [inline vowpalwabbit/gen_cs_example.h]: Is there a reason why you sometimes use safe_probability() and sometimes not?

Comment (marco-rossi29) [reply vowpalwabbit/gen_cs_example.h]: I've keep everything as is. Not editing the use of safe_probability()

**Comment by JohnLangford** [comment]
Merged, thanks :-)

**Pull request merged**